plugins/inventory/aws_ec2.py

```diff
@@ -505,7 +510,14 @@ def _get_hostname(self, instance, hostnames):
 
         hostname = None
         for preference in hostnames:
-            if 'tag' in preference:
+            if isinstance(preference, dict):
+                if 'name' not in preference:
+                    raise AnsibleError("A 'name' key must be defined in a hostnames dictionary.")
+                hostname = self._get_hostname(instance, [preference["name"]])
+                separator = preference.get("separator", "_")
+                if 'prefix' in preference:
+                    hostname = self._get_hostname(instance, [preference["prefix"]]) + separator + hostname
```

Comment:
If `_get_hostname` return `None` (e.g: tag is missing on a VM), I get:
```
unsupported operand
type(s) for +: 'NoneType' and 'AnsibleUnicode'
```
So it would be better to do it in two steps:

```python
if 'prefix' in preference and self._get_hostname(instance, [preference["prefix"]]):
    hostname = self._get_hostname(instance, [preference["prefix"]]) + separator + hostname
```

Reply:
Done, added a leading `hostname and` to the condition

Reply:
This does not resolve my problem, `self._get_hostname()` can still return `None` if there is no match. I use this patch: 

```
diff --git a/plugins/inventory/aws_ec2.py b/plugins/inventory/aws_ec2.py
index 29fa9a1..2d6568f 100644
--- a/plugins/inventory/aws_ec2.py
+++ b/plugins/inventory/aws_ec2.py
@@ -514,10 +514,9 @@ class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
                 hostname = self._get_hostname(instance, [preference["name"]])
                 separator = preference.get("separator", "_")
                 if hostname and 'prefix' in preference:
-                    hostname = self._get_hostname(instance, [preference["prefix"]]) + separator + hostname
-            elif 'tag' in preference:
-                if not preference.startswith('tag:'):
-                    raise AnsibleError("To name a host by tags name_value, use 'tag:name=value'.")
+                    hostname_from_prefix = self._get_hostname(instance, [preference["prefix"]]) or ''
+                    hostname = hostname_from_prefix + separator + hostname
+            elif preference.startswith('tag:'):
                 hostname = self._get_tag_hostname(preference, instance)
             else:
                 hostname = self._get_boto_attr_chain(preference, instance)
```

With this configuration:

```yaml
hostnames:                                                                                    
  - name: 'tag:Name=aws:cloudformation:stack-id'                                              
  - name: 'private-ip-address'                                                                
    separator: '_'                                                                            
    prefix: 'tag:Name'
```

I get:

- _10.0.0.43 <---- There is no matching tag for this one, so the prefix is ''
- _10.0.234.96 <---- the same
- ansible-operator-rhel-to-rhel-responder-instance_192.168.0.10
- cvpn-initiator-instance_10.0.0.10
            "cvpn-responder-instance_192.168.0.10"

My question at this stage, should we skip these two first instances, or keep them in the list?

Reply:
Sometimes, at the very early stage of initialization, all instance tags may not have been set.
We have solved this on our case by having a wait loop in the cloud-init.

I don't know if it is useful to have such inventory names with no prefix, though.
I propose to skip them, and if someone has a use-case, we could in the future add them.